### PR TITLE
Add `ignoreFiles` option

### PR DESCRIPTION
**Pull request by jridgewell**
This is very similar to the `gitignore` option, except it allows globing for any compatible ignore file (eg, `.prettierignore` or `.eslintignore`). AMP had a recent need for this with our `prettier` integration, because it doesn't respect its ignore file when using the API directly.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -167,15 +180,20 @@ module.exports.stream = (patterns, options) => {
 	const filterStream = new FilterStream(p => !filter(p));
 	const uniqueStream = new UniqueStream();
 
-	return merge2(tasks.map(task => fastGlob.stream(task.pattern, task.options)))
+	return merge2(
+		tasks.map(task => fastGlob.stream(task.pattern, task.options))
+	)
```
Comment (sindresorhus) [inline index.js:185]: Don't do unrelated changes

File: index.d.ts [hunk 2]
```diff
@@ -39,6 +39,13 @@ declare namespace globby {
 		@default false
 		*/
 		readonly gitignore?: boolean;
+
+		/**
+		Glob pattern to look for ignore files, which are then used to ignore globbed files.
```
Comment (sindresorhus) [inline index.d.ts:44]: The definition of what "ignore files" are is not clear. The docs need to be expanded. There are also many unanswered questions: How does this relate to `gitignore` option. Does it work in addition to it, instead of it, etc.

You also need to add docs to the readme.

File: index.d.ts [hunk 3]
```diff
@@ -39,6 +39,13 @@ declare namespace globby {
...
+
+		@default undefined
```
Comment (sindresorhus) [inline index.d.ts:46]: This is not correct.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.js (hunk 1 above)
Comment (jridgewell) [reply index.js:185]: Undone.

File: index.d.ts (hunk 3 above)
Comment (jridgewell) [reply index.d.ts:46]: Can you explain? If `ignoreFiles` isn't specified, we don't do anything with the value. Would you rather this say "empty string", or something else?

File: index.d.ts (hunk 2 above)
Comment (jridgewell) [reply index.d.ts:44]: Added a quick explanation here, and a larger explanation to the readme.

File: index.js [hunk 4]
```diff
@@ -91,17 +91,25 @@ const globDirs = (task, fn) => {
...
 };
 
 const globToTask = task => glob => {
 	const {options} = task;
-	if (options.ignore && Array.isArray(options.ignore) && options.expandDirectories) {
+	if (
+		options.ignore &&
+		Array.isArray(options.ignore) &&
+		options.expandDirectories
+	) {
```
Comment (sindresorhus) [inline index.js:112]: Please don't do unrelated changes. Undo all unrelated changes.

**Comment by sindresorhus** [comment]
You need to update package.json when renaming a file.

File: index.d.ts [hunk 5]
```diff
@@ -39,6 +39,14 @@ declare namespace globby {
 		@default false
 		*/
 		readonly gitignore?: boolean;
+
+		/**
+		Glob pattern to look for ignore files, which are then used to ignore globbed files.
+		This value is unused if gitignore option is used, since gitignore takes precedence.
```
Comment (sindresorhus) [inline index.d.ts:45]: This is ambiguous, setting `gitignore: false` is technically *it being used*. It should explicitly say that setting `gitignore: true` overrides this option.

File: readme.md [hunk 6]
```diff
@@ -87,7 +87,14 @@ Note that if you set this option to `false`, you won't get back matched director
...
 
-Respect ignore patterns in `.gitignore` files that apply to the globbed files.
+Respect ignore patterns in `.gitignore` files that apply to the globbed files. This option takes precedence over the `ignoreFiles` option, and explicitly looks for files using a `**/.gitignore` glob.
+
+##### ignoreFiles
+
+Type: `string`\
+Default: `undefined`
+
+A glob pattern to look for ignore files, which are then used to ignore globbed files. This is a more generic form of the `gitignore` option, allowing you to find ignore files with a [compatible syntax](http://git-scm.com/docs/gitignore). For instance, this is works with Babel's `.babelignore`, Prettier's `.prettierignore`, or ESLint's `.eslintignore` files.
```
Comment (sindresorhus) [inline readme.md:97]: Readme docs and index.d.ts docs should be in sync.

File: index.d.ts (hunk 3 above)
Comment (sindresorhus) [reply index.d.ts:47]: Sorry, I don't remember. I should have been clearer.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: readme.md [hunk 7]
```diff
@@ -10,7 +10,7 @@ Based on [`fast-glob`](https://github.com/mrmlnc/fast-glob) but adds a bunch of
 - Multiple patterns
 - Negated patterns: `['foo*', '!foobar']`
 - Expands directories: `foo` → `foo/**/*`
-- Supports `.gitignore`
+- Supports `.gitignore`, or any similar ignore file system.
```
Comment (sindresorhus) [inline readme.md:13]: This reads like "file system", not "ignore file  .. system". Use a different word than system.

File: index.js (hunk 4 above)
Comment (jridgewell) [reply index.js]: Undone. It's the dang autoformatter.

File: index.d.ts (hunk 5 above)
Comment (jridgewell) [reply index.d.ts]: Done.

File: readme.md (hunk 6 above)
Comment (jridgewell) [reply readme.md:99]: Synced.

File: readme.md (hunk 7 above)
Comment (jridgewell) [reply readme.md]: Changed to "configuration", it was the best I could think of.

**Review by jridgewell — COMMENTED** [review]
Addressed all comments, and made sure to undo any autoformatting (sorry about that).

**Comment by weyert** [comment]
How can I help getting this PR merged into the project?

**Comment by jridgewell** [comment]
I've rebased, this just requires a review.

File: index.d.ts [hunk 8]
```diff
@@ -42,11 +42,20 @@ export interface Options extends FastGlobOptionsWithoutCwd {
...
 	*/
 	readonly gitignore?: boolean;
 
+	/**
+	A glob pattern to look for ignore files, which are then used to ignore globbed files. This is a more generic form of the `gitignore` option, allowing you to find ignore files with a [compatible syntax](http://git-scm.com/docs/gitignore). For instance, this works with Babel's `.babelignore`, Prettier's `.prettierignore`, or ESLint's `.eslintignore` files.
+	This value is unused if `{ gitignore: true }`, since `gitignore` takes precedence.
+
+	@default undefined
+	*/
+	readonly ignoreFiles?: string;
```
Comment (fisker) [inline index.d.ts:57]: Why not take `string[] | sting` for easier use.

File: ignore.js [hunk 9]
```diff
@@ -77,21 +77,21 @@ const normalizeOptions = (options = {}) => ({
...
 
 	return getIsIgnoredPredicate(ignores, cwd);
 };
 
-export const isGitIgnoredSync = options => {
+export const isIgnoredSync = (pattern, options) => {
 	const {cwd} = normalizeOptions(options);
 
-	const paths = fastGlob.sync('**/.gitignore', {cwd, ...gitignoreGlobOptions});
+	const paths = fastGlob.sync(pattern, {cwd, ...gitignoreGlobOptions});
```
Comment (fisker) [inline ignore.js:94]: Maybe we should add `dot: true` to the option, when pattern is `*ignore`, not sure, need test.

File: index.d.ts [hunk 10]
```diff
@@ -42,11 +42,20 @@ export interface Options extends FastGlobOptionsWithoutCwd {
 
 	/**
 	Respect ignore patterns in `.gitignore` files that apply to the globbed files.
+	This option takes precedence over the `ignoreFiles` option.
```
Comment (fisker) [inline index.d.ts:45]: This don't have to "takes precedence", we can concat `**/.gitignore` with `ignoreFiles`, so they can use together.

File: index.d.ts (hunk 8 above)
Comment (jridgewell) [reply index.d.ts:57]: Done.

File: index.d.ts (hunk 10 above)
Comment (jridgewell) [reply index.d.ts:45]: Done.

File: ignore.js (hunk 9 above)
Comment (jridgewell) [reply ignore.js:94]: > Maybe we should add dot: true to the option, when pattern is `*ignore`

Sorry, I don't understand the suggestion.

> need test.

Added.

Comment (fisker) [reply ignore.js:94]: I mean, if user use `{ignoreFiles: '*ignore'}`, can this glob files starts with `.`? See https://github.com/mrmlnc/fast-glob#dot

File: index.js [hunk 11]
```diff
@@ -14,7 +14,7 @@ const assertPatternsInput = patterns => {
 };
 
 const toPatternsArray = patterns => {
-	patterns = [...new Set([patterns].flat())];
+	patterns = [...new Set([patterns].flat(Infinity))];
```
Comment (fisker) [inline index.js:17]: Instead of flat infinity, why not `toPatternsArray(options.ignoreFiles)` then push `**/.gitignore`? Changing this will make `patterns` for `globby()` behave differently.

File: ignore.js (hunk 9 above)
Comment (jridgewell) [reply ignore.js:94]: I personally don't find that necessary, and it makes the configuration a little complicated (what if the user didn't want to use dotfiles?). But I'll leave the decision to you.

File: index.js (hunk 11 above)
Comment (jridgewell) [reply index.js:17]: Done.

File: ignore.js (hunk 9 above)
Comment (fisker) [reply ignore.js:94]: I'm not a maintainer :)

File: index.js [hunk 12]
```diff
@@ -52,13 +52,32 @@ const normalizeOptions = (options = {}) => {
...
 
-const getFilterSync = options => createFilterFunction(
-	options.gitignore && isGitIgnoredSync({cwd: options.cwd}),
-);
+	const patterns = toPatternsArray(options.ignoreFiles || []);
+	if (options.gitignore) {
+		patterns.push('**/.gitignore');
+	}
+
+	return toPatternsArray(patterns);
```
Comment (fisker) [inline index.js:65]:
```suggestion
	return patterns;
```

File: index.js [hunk 13]
```diff
@@ -200,7 +219,6 @@ export const isDynamicPattern = normalizeArgumentsSync(
 export const generateGlobTasks = normalizeArguments(generateTasks);
 export const generateGlobTasksSync = normalizeArgumentsSync(generateTasksSync);
 
-export {
-	isGitIgnored,
-	isGitIgnoredSync,
-} from './gitignore.js';
+export {isIgnored, isIgnoredSync} from './ignore.js';
```
Comment (fisker) [inline index.js:222]: I'm not sure if we want to expose these, if we do, maybe we need a better name, `isIgnoredByIgnoreFiles`?

File: index.js (hunk 12 above)
Comment (jridgewell) [reply index.js]: Ha.

File: index.d.ts (hunk 10 above)
Comment (fisker) [reply index.d.ts]: The content in readme.md file didn't fix.

**Comment by jridgewell** [comment]
This has gotten another merge conflict. I'm not interested in doing more rounds. If you want to take this up, I release my changes as public domain.

**Comment by fisker** [comment]
I can continue work on this.
